duktape/src/main/jni/DuktapeContext.cpp

```diff
@@ -228,44 +192,26 @@ void DuktapeContext::bind(JNIEnv *env, jstring name, jobject object, jobjectArra
   duk_pop(m_context);
 }
 
-jobject DuktapeContext::proxy(JNIEnv* env, jstring name, jobjectArray methods) {
-  // TODO: lookup and attach to object 'name' in the JS VM, set up call wrappers for methods given.
-  return name;
+jlong DuktapeContext::proxy(JNIEnv* env, jstring name, jobjectArray methods) {
+  const JString instanceName(env, name);
+  jlong index = static_cast<jlong>(m_proxiedObjects.size());
+  try {
+    m_proxiedObjects.push_back(new JavaScriptObject(env, m_context, instanceName.str(), methods));
+  } catch (const std::invalid_argument& e) {
+    queueIllegalArgumentException(env, e.what());
+  } catch (const std::runtime_error& e) {
+    queueDuktapeException(env, e.what());
+  }
+  return index;
 }
 
-jobject DuktapeContext::call(JNIEnv *env, jobject target, jobject method, jobjectArray args) {
-  duk_push_global_object(m_context);
-  // TODO: attach to something stronger than the name in proxy(), and use that here.
-  const JString instanceName(env, static_cast<jstring>(target));
-  if (!duk_get_prop_string(m_context, -1, instanceName)) {
-    duk_pop(m_context);
-    const jclass illegalArgumentException = env->FindClass("java/lang/IllegalArgumentException");
-    const std::string message = "A global object called " + instanceName.str() + " was not found";
-    env->ThrowNew(illegalArgumentException, message.c_str());
+jobject DuktapeContext::call(JNIEnv *env, jlong target, jobject method, jobjectArray args) {
+  JavaScriptObject *object = target < m_proxiedObjects.size()
```

Comment:
It's implicit since `m_proxiedObjects.size()` is unsigned and `target` will be changed to unsigned for the compare, but I'll add the check to make it more apparent.